State: Include updates in sites, update sites/updates reducer

tyxla opened the pull request:
Currently, `/me/sites/` gives us `updates` for Jetpack sites. However, we simply disregard it right now.

This PR adds `updates` to the sites schema, so it will be included in the Redux site objects. Also, it removes the entire `sites/updates` tree and the `QuerySiteUpdates` query component, as they'll be unnecessary now that we'll contain updates information in `sites`. We can safely remove this subtree and this component, because they're not used anywhere yet.

To test:
* Checkout this branch
* Verify all sites tests pass:

```
npm run test-client client/state/sites/
```

gwwar commented on client/state/sites/schema.js:
```diff
@@ -57,7 +57,8 @@ export const sitesSchema = {
 				},
-				single_user_site: { type: 'boolean' }
+				single_user_site: { type: 'boolean' },
+				updates: { type: 'object' },
```
Is this guaranteed to be defined?

tyxla replied:
@gwwar It depends on the type of site really:

* ✅  for single-site Jetpack sites
* ✅  for the main site of Jetpack multisite networks
* ❌  for sub-sites of Jetpack multisite networks
* ❌  for WordPress.com sites

Which is why it is not a required attribute. But we should be adding it to the sites schema, because it's already used here and there (for example, see https://github.com/Automattic/wp-calypso/blob/master/client/lib/sites-list/list.js#L578 or https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/site-indicator/site-indicator.jsx#L30).

gwwar replied:
To clarify, unless we specify `additionalProperties: false` in the schema, additional unexpected properties are tolerated during the schema check. So in this case if we actually added this, it'd need to be `updates: type:[ 'object', 'null' ]` otherwise we'd throw out any persisted site data for normal wpcom sites.

gwwar replied:
Tested this, and looks like it's ok here with `undefined`. Just be careful with some of our endpoints, at times they may return different types or `null` for optional properties.

gwwar commented:
I think it's fine to leave the sub-tree here, our client state doesn't need to mirror api responses. We can instead add the relevant events to the updates reducer. For example a the currentUser's currency code can be fetched from multiple endpoints, so in the reducer this looks like: https://github.com/Automattic/wp-calypso/blob/master/client/state/current-user/reducer.js#L48

cc @retrofox

tyxla commented:
@gwwar on second thought that makes sense. Decoupling anything from the `site` object could help with performance, and indeed it could be helpful to fetch fresh updates quickly, instead of waiting for `/me/sites` to load. I'll update the PR to NOT remove `updates` stuff, but rather update it when sites arrive.

tyxla commented:
I've updated the PR to use the `sites/updates` state tree instead of deleting it, but I found some crucial errors with the current implementation that I also resolved here. So this PR now handles the following:

* Adds `updates` to the sites schema, so we can use it where necessary.
* Updates `sites/updates` (if `updates` is present) when we:
  * Receive new sites
  * Receive a new site
  * Update sites
* Fixed a bug with the current sites updates schema - allowed props were incorrectly wrapped in an `items` property.
* Fixed `sites/updates` reducer to store `requesting` on a per-site basis. Selectors are written to expect that, but reducer was not doing it.
* Fixed `sites/updates` reducer to store `errors` on a per-site basis. 
* Added tests for the `sites/updates` reducer.

I realize this PR has grown too large, but this is because `sites/updates` currently has no tests.

@gwwar it would be awesome if you could take another look. Also @retrofox's feedback would be valuable, as he worked on `sites/updates`.

tyxla commented:
Also, I've opened PRs that add missing tests for:

* Action creators - #13077 
* Selectors - #13078
* Utils - #13079

mcsf commented on client/state/sites/updates/reducer.js:
```diff
@@ -2,60 +2,61 @@
+	return reduce( sites, ( memo, site ) => {
+		if ( memo === state ) {
+			memo = { ...state };
+		}
```
I assume this copy is made here rather than in `reduce`'s last argument so that we don't create another instance of `state` if `sites` is empty? If so, this will not prevent the following case:
- `sites` is not empty
- `sites` does not contain any `site` where `site.updates`
- return value is now a different instance, even though it is equal to previous `state`

I would suggest maybe something like:

```js
const receiveUpdatesForSites = ( state, sites ) => {
  const updatedSites = sites.filter( ( site ) => site.updates );
  return isEmpty( updatedSites )
    ? state
    : sites.reduce( ( newState, site ) => {
        newState[ site.ID ] = site.updates;
      }, { ...state } );
}
```

tyxla replied:
Good point! :+1:

mcsf commented on client/state/sites/updates/reducer.js:
```diff
@@ -2,60 +2,61 @@
+export const items = createReducer(
+	{},
+	{
+		[ SITE_UPDATES_RECEIVE ]: ( state, { siteId, updates } ) => Object.assign( {}, state, { [ siteId ]: updates } ),
```
Object spread notation for brevity? `{ ...state, [ siteId ]: updates }`

mcsf commented on client/state/sites/updates/reducer.js:
```diff
@@ -2,60 +2,61 @@
+	[ SITE_UPDATES_REQUEST ]: ( state, { siteId } ) => ( { ...state, [ siteId ]: true } ),
+	[ SITE_UPDATES_REQUEST_SUCCESS ]: ( state, { siteId } ) => ( { ...state, [ siteId ]: false } ),
+	[ SITE_UPDATES_REQUEST_FAILURE ]: ( state, { siteId } ) => ( { ...state, [ siteId ]: false } ),
+} );
```
These could maybe benefit from [`keyedReducer`](https://github.com/Automattic/wp-calypso/blob/50d131b90defd57c4382a0b97bb96caae3cc6517/client/state/utils.js#L77), but don't let that keep you from merging.

tyxla replied:
Ah, haven't used `keyedReducer` so far. That's very handy, thanks!

tyxla commented:
Now let's get someone else to review that one 😉

gwwar commented on client/state/sites/updates/reducer.js:
```diff
@@ -2,60 +2,57 @@
+	[ SITE_UPDATES_REQUEST ]: stubFalse,
+	[ SITE_UPDATES_REQUEST_SUCCESS ]: stubFalse,
+	[ SITE_UPDATES_REQUEST_FAILURE ]: stubTrue,
+} ) );
```
I wonder if we need this anymore? Any error notices would probably be better handled in the notice middleware.

tyxla replied:
Agree - it's not necessary right now. I guess I was hesitant to remove it as we refactored the entire reducer. But we can remove it in a subsequent PR, right?

gwwar replied:
Yup future refactors 👍

gwwar commented on client/state/sites/updates/reducer.js:
```diff
@@ -2,60 +2,57 @@
+	[ SITE_UPDATES_REQUEST ]: stubTrue,
+	[ SITE_UPDATES_REQUEST_SUCCESS ]: stubFalse,
+	[ SITE_UPDATES_REQUEST_FAILURE ]: stubFalse,
+} ) );
```
Once we move this to the data layer this reducer will be unnecessary, but no need to do it here since we're a bit time constrained this week.
